### ISSUE TITLE
Require data to match coll type

### ISSUE THREAD

**Issue by brjann** [metosin/spec-tools#237]
It does not seem to be possible to require data to match the type of collection in a data spec.

For example both `(s/valid? (ds/spec ::x #{int?}) [1 2 3])` and `(s/valid? (ds/spec ::x [int?]) #{1 2 3})` return true. I have been bitten by this where I accidentally passed a list instead of a set to a function, which passed the spec check and since `contains?` accepts list, no error occured. 

Is it possible to require an exact match on collection type?

**Comment by ikitommi** [comment]
sounds like a bug. PR welcome.

**Comment by brjann** [comment]
I wish I had enough Clojure / spec skills to submit a PR on this :-)

**Comment by brjann** [comment]
While 
`(s/valid? (ds/spec ::x {int? int?}) #{1 2})` returns false, 
`(s/valid? (ds/spec ::x {int? int?}) #{[1 2]})` returns true

So it seems that one must be able to convert the value to its correct form for the spec check to succeed

**Comment by wandersoncferreira** [comment]
@brjann there is some explanation in the mentioned PR above to fix the issue you reported.

**Comment by brjann** [comment]
@wandersoncferreira wow, thanks! Does this also make `(s/valid? (ds/spec ::x {int? int?}) #{[1 2]})` returen false?

**Comment by wandersoncferreira** [comment]
Unfortunately not @brjann , it seems related but it is not the same error. From some investigation, seems like the code performing the destructuring of `#{[1 2]}` is treating it the same way as `#{1 2}`. I am trying to look where this is done.

**Comment by wandersoncferreira** [comment]
Which is interesting is that `(s/valid? (ds/spec ::x {int? int?}) #{[1 2 3]})` returns false as expected. This error is only happening with exact 2 elements inside the array.

**Comment by brjann** [comment]
I would guess that ´[k v]´ is identified as a key value pair. `(s/valid? (ds/spec ::x {int? int?}) [[1 2] [2 1]])` is also valid, while `(s/valid? (ds/spec ::x {int? int?}) [[1 2] [2 1 3]])` is not.
Also, `(s/conform (ds/spec ::x {int? int?}) [[1 2] [2 1]])` => `{1 2 2 1}`

**Comment by wandersoncferreira** [comment]
Yes, I got the whole picture now.

When the `map-of-spec` is called to create our data spec, we rely on `clojure.spec.alpha/every-impl` to create it. However, this function has the following logic to perform the dispatch for each value when you ask to `conform-all` values in the map: https://github.com/clojure/spec.alpha/blob/master/src/main/clojure/clojure/spec/alpha.clj#L1298

And the problem lies in the `(seq x)` call. `x` is our data in this case `#{[1 2 3 4]}` . However, take a look at the destructuring below (simplified version of what the `every-impl` does):

```clj
;; the value of v
(let [x #{[1 2 3]}
      [v & vs :as vseq] (seq x)]
  v) ;; => [1 2 3]

;;; where v is used
(let [x #{[1 2 3]}
      [v & vs :as vseq] (seq x)]
  (conform* spec v))
```

And what is `spec`? Looking at `impl/map-of-spec`, spec is checking for a key-value `tuple`. And `seq` just got rid off our `#{}` literal. Then when the number of itens inside the vector is even, `(conform* spec v)` returns true and everything is fine.

Now, I'm not sure how can we approach to fix the issue. I looked under the hood of `s/map-of` which seems to do this just fine, however the implementation to some of the parts, e.g. `cpred` parameter, is a lot more complex than the current `coll?` call spec-tools is using right now. 

I will continue to look at this problem when I get more time.
